# Homework 2

The goal of this Homework is to write Python code to answer each of the questions below. You should be able to answer these questions using Pandas only. You should not need to import any other packages. You should not use Excel or any other means to answer these questions. **Since questions have subquestions, it would be helpful if you use comments to clearly indicate which question you are answering**.

## Introduction
The Centers for Medicare & Medicaid Services ([CMS](cms.gov)) is a federal agency within the United States Department of Health and Human Services (HHS) that administers the **Medicare program** and works in partnership with state governments to administer Medicaid, the Children's Health Insurance Program (CHIP), and health insurance portability standards.
 - **Medicare** is a national health insurance program that provides health insurance for Americans aged 65 and over

The **Medicare Part D** program, also called the Medicare prescription drug benefit, is an optional United States federal-government program to help Medicare beneficiaries pay for self-administered prescription drugs through prescription drug insurance premiums.

CMS maintains a lot of [public info](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/Part-D-Prescriber) on prescription drugs that are prescribed by individual physicians and other health care **providers** and paid for under the **Medicare Part D** Prescription Drug Program.

I downloaded the **2017 Provider Summary table** and sampled it down from 1.1 million records (360 MB) to roughly 6000 records (2MB). If you would like to check out the original dataset, you can download it from this site https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/PartD2017   

But for this assignment, you need to use the provided downsampled dataset `part_d_drug_prescriptions_sample2.csv`. If you would like, the data dictionary is available [here](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/Downloads/Prescriber_Methods.pdf). The applicable section is Table 2 in the Appendix.


In [8]:
import pandas as pd
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_rows', 200)


## Question 1

Use Pandas to import **part_d_drug_prescriptions_sample2.csv** into a DataFrame, named `part_d`. How many rows and colums does the dataset have?

In [9]:
part_d = pd.DataFrame(pd.read_csv('part_d_drug_prescriptions_sample2.csv',header=0))
part_d

npi nppes_provider_last_org_name nppes_provider_first_name  \
0     1003000126                    ENKESHAFI                   ARDALAN   
1     1003012022                      CUDZILO                     COREY   
2     1003020041                       BAXTER                  MARGARET   
3     1003031212                       KAYALI                      FADI   
4     1003043175                       BENSON                   REBECCA   
...          ...                          ...                       ...   
5810  1992949150                         WANG                       AMY   
5811  1992961429                       PARKER                   MELINDA   
5812  1992969141                      BURKART                      NORA   
5813  1992980767                 TANAWUTTIWAT                  TANYANAN   
5814  1992994446                       PATETE                   MICHAEL   

     nppes_provider_mi nppes_credentials nppes_provider_gender  \
0                  NaN              M.D.                     M   
1                  NaN              M.D.                     M   
2                  NaN              M.D.                     F   
3                  NaN              M.D.                     M   
4                    L            D.D.S.                     F   
...                ...               ...                   ...   
5810                 M                MD                     F   
5811                 S              PA-C                     F   
5812                 E              M.D.                     F   
5813               NaN              M.D.                     F   
5814                 L          MD, FACS                     M   

     nppes_entity_code  nppes_provider_street1 nppes_provider_street2  \
0                    I            900 SETON DR                    NaN   
1                    I     2240 SUTHERLAND AVE              SUITE 103   
2                    I       1100 LONG POND RD              SUITE 250   
3                    I      600 N CATTLEMEN RD              SUITE 200   
4                    I      6601 LYNDALE AVE S                    NaN   
...                ...                     ...                    ...   
5810                 I  3130 N COUNTY ROAD 25A                    NaN   
5811                 I          3209 S 23RD ST                STE 340   
5812                 I     1805 HENNEPIN AVE N                    NaN   
5813                 I         2500 N STATE ST                    NaN   
5814                 I          213 PALERMO PL                    NaN   

     nppes_provider_city  ...  beneficiary_male_count  \
0             CUMBERLAND  ...                    92.0   
1              KNOXVILLE  ...                    35.0   
2              ROCHESTER  ...                   221.0   
3               SARASOTA  ...                   140.0   
4              RICHFIELD  ...                     NaN   
...                  ...  ...                     ...   
5810                TROY  ...                    73.0   
5811              TACOMA  ...                    59.0   
5812             GLENCOE  ...                    11.0   
5813             JACKSON  ...                    84.0   
5814              VENICE  ...                   472.0   

      beneficiary_race_white_count beneficiary_race_black_count  \
0                            220.0                         14.0   
1                              NaN                          NaN   
2                            544.0                          NaN   
3                            249.0                          NaN   
4                              NaN                          0.0   
...                            ...                          ...   
5810                         168.0                         24.0   
5811                           NaN                          0.0   
5812                           NaN                          0.0   
5813                           NaN                         79.0   
5814                

In [10]:
part_d.shape

(5815, 84)

## Question 2
This question has two parts:
1. Display the first 6 rows of `part_d`.
2. Display the last 6 rows of `part_d`.

In [11]:
part_d.head(6)

npi nppes_provider_last_org_name nppes_provider_first_name  \
0  1003000126                    ENKESHAFI                   ARDALAN   
1  1003012022                      CUDZILO                     COREY   
2  1003020041                       BAXTER                  MARGARET   
3  1003031212                       KAYALI                      FADI   
4  1003043175                       BENSON                   REBECCA   
5  1003054206                       SEILER                    WARREN   

  nppes_provider_mi nppes_credentials nppes_provider_gender nppes_entity_code  \
0               NaN              M.D.                     M                 I   
1               NaN              M.D.                     M                 I   
2               NaN              M.D.                     F                 I   
3               NaN              M.D.                     M                 I   
4                 L            D.D.S.                     F                 I   
5                 B                MD                     M                 I   

  nppes_provider_street1 nppes_provider_street2 nppes_provider_city  ...  \
0           900 SETON DR                    NaN          CUMBERLAND  ...   
1    2240 SUTHERLAND AVE              SUITE 103           KNOXVILLE  ...   
2      1100 LONG POND RD              SUITE 250           ROCHESTER  ...   
3     600 N CATTLEMEN RD              SUITE 200            SARASOTA  ...   
4     6601 LYNDALE AVE S                    NaN           RICHFIELD  ...   
5         2700 ROGERS DR              SUITE 101            HOMEWOOD  ...   

   beneficiary_male_count  beneficiary_race_white_count  \
0                    92.0                         220.0   
1                    35.0                           NaN   
2                   221.0                         544.0   
3                   140.0                         249.0   
4                     NaN                           NaN   
5                     NaN                           NaN   

  beneficiary_race_black_count beneficiary_race_asian_pi_count  \
0                         14.0                             0.0   
1                          NaN                             0.0   
2                          NaN                             NaN   
3                          NaN                             NaN   
4                          0.0                             0.0   
5                          NaN                             NaN   

  beneficiary_race_hispanic_count beneficiary_race_nat_ind_count  \
0                             0.0                            0.0   
1                             NaN                            0.0   
2                            17.0                            0.0   
3                             NaN                            0.0   
4                             0.0                            0.0   
5                             NaN                            NaN   

  beneficiary_race_other_count  beneficiary_nondual_count  \
0                          0.0                      143.0   
1                          0.0                       49.0   
2                         26.0                      515.0   
3                          NaN                      229.0   
4                          NaN                       20.0   
5                          NaN                        NaN   

   beneficiary_dual_count  beneficiary_average_risk_score  
0                    91.0                          2.1685  
1                    33.0                          2.5022  
2                    95.0                          1.0164  
3                    45.0                          2.4311  
4                     0.0                          1.0910  
5                     NaN                             NaN  

[6 rows x 84 columns]

In [12]:
part_d.tail(6)

npi nppes_provider_last_org_name nppes_provider_first_name  \
5809  1992936330                       MANLEY               CHRISTOPHER   
5810  1992949150                         WANG                       AMY   
5811  1992961429                       PARKER                   MELINDA   
5812  1992969141                      BURKART                      NORA   
5813  1992980767                 TANAWUTTIWAT                  TANYANAN   
5814  1992994446                       PATETE                   MICHAEL   

     nppes_provider_mi nppes_credentials nppes_provider_gender  \
5809                 J                MD                     M   
5810                 M                MD                     F   
5811                 S              PA-C                     F   
5812                 E              M.D.                     F   
5813               NaN              M.D.                     F   
5814                 L          MD, FACS                     M   

     nppes_entity_code  nppes_provider_street1 nppes_provider_street2  \
5809                 I         333 COTTMAN AVE                    NaN   
5810                 I  3130 N COUNTY ROAD 25A                    NaN   
5811                 I          3209 S 23RD ST                STE 340   
5812                 I     1805 HENNEPIN AVE N                    NaN   
5813                 I         2500 N STATE ST                    NaN   
5814                 I          213 PALERMO PL                    NaN   

     nppes_provider_city  ...  beneficiary_male_count  \
5809        PHILADELPHIA  ...                    25.0   
5810                TROY  ...                    73.0   
5811              TACOMA  ...                    59.0   
5812             GLENCOE  ...                    11.0   
5813             JACKSON  ...                    84.0   
5814              VENICE  ...                   472.0   

      beneficiary_race_white_count beneficiary_race_black_count  \
5809                           NaN                          NaN   
5810                         168.0                         24.0   
5811                           NaN                          0.0   
5812                           NaN                          0.0   
5813                           NaN                         79.0   
5814                         951.0                          NaN   

     beneficiary_race_asian_pi_count beneficiary_race_hispanic_count  \
5809                             NaN                             0.0   
5810                             NaN                             NaN   
5811                             0.0                             0.0   
5812                             0.0                             NaN   
5813                             NaN                             NaN   
5814                             NaN                             NaN   

     beneficiary_race_nat_ind_count beneficiary_race_other_count  \
5809                            0.0                          NaN   
5810                            0.0                          NaN   
5811                            0.0                          NaN   
5812                            0.0                          0.0   
5813                            NaN                          0.0   
5814                            NaN                         19.0   

      beneficiary_nondual_count  beneficiary_dual_count  \
5809                        NaN                     NaN   
5810                      146.0                    58.0   
5811                        NaN                     NaN   
5812                        NaN                     NaN   
5813                       75.0                    78.0   
5814                      948.0                    35.0   

      beneficiary_average_risk_score  
5809                          2.5471  
5810                          2.0090  
5811                          0.9592  
5812                          1.3618  
5813                          2.0090  
5814        

## Question 3
This question has 4 parts:
1. Use the `info()` function to examine the DataFrame from a high level. Take a moment to look at the number of non-null values and the datatypes of each column. 
2. How many non-null values does the `npi` column have? (Use the output of the `info()` function to answer this question.)
3. What is the data type of the `npi` column? (Use the output of the `info()` function to answer this question.)
4. The purpose of this question is to understand the data you are looking at. The `npi` stands for National Provider Identifier. Please provide a one or two sentence description of the NPI - you may use a web search (google/wikipedia/etc...) to answer this question. 

In [13]:
part_d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5815 entries, 0 to 5814
Data columns (total 84 columns):
npi                                 5815 non-null int64
nppes_provider_last_org_name        5815 non-null object
nppes_provider_first_name           5815 non-null object
nppes_provider_mi                   4074 non-null object
nppes_credentials                   5461 non-null object
nppes_provider_gender               5815 non-null object
nppes_entity_code                   5815 non-null object
nppes_provider_street1              5815 non-null object
nppes_provider_street2              2107 non-null object
nppes_provider_city                 5815 non-null object
nppes_provider_zip5                 5815 non-null float64
nppes_provider_zip4                 5031 non-null float64
nppes_provider_state                5815 non-null object
nppes_provider_country              5815 non-null object
specialty_description               5815 non-null object
description_flag                    5

Answer to Question 3
This question has 4 parts:
1. Command has been called
2. Through observation, there are 5815 non-null values
3. int64
4. A National Provider Identifier (NPI) is a unique 10-digit identification number issued to health care providers in the United States by the Centers for Medicare and Medicaid Services (CMS). The NPI has replaced the unique physician identification number (UPIN) as the required identifier for Medicare services, and is used by other payers, including commercial healthcare insurers. 

## Question 4
Take a look at the `nppes_provider_state` column. Compute the number of unique values in this column using at least two different methods.

In [14]:
data = part_d['nppes_provider_state']
data.unique().shape[0]

55

In [15]:
part_d['nppes_provider_state'].value_counts().shape[0]

55

## Question 5
Let's take a look at the `specialty_description` column. Use the `value_counts()` method to determine the TOP 5 specialties in the dataset as well as their frequencies.

In [16]:
data = part_d['specialty_description']
data

0                        Internal Medicine
1                        Pulmonary Disease
2                          Family Practice
3                      Hematology-Oncology
4                                  Dentist
                       ...                
5810                     Internal Medicine
5811                   Physician Assistant
5812                       General Surgery
5813    Clinical Cardiac Electrophysiology
5814                        Otolaryngology
Name: specialty_description, Length: 5815, dtype: object

In [17]:
data.value_counts().head(5)

Nurse Practitioner     785
Dentist                647
Internal Medicine      626
Family Practice        565
Physician Assistant    446
Name: specialty_description, dtype: int64

## Question 6
This question has four parts:
1. Define a variable `N` equal to the number of rows in the DataFrame, using the DataFrame `len` or `shape` methods.
2. Append the `cumsum()` function to the output of the `value_counts()` function on the `specialty_description` field  and divide the output by `N` to obtain the cumulative ratio to the total count. 
3. Create a boolean variable (eg a mask) which resolves to `True` if the cumulative sum is less than 75%, else `False`.
4. Use this boolean mask to filter the output of `value_counts()` to identify the columns that contribute to the first 75% of all specialities.

In [18]:
N = part_d.shape[0]
print(N)
data = part_d['specialty_description'].value_counts()
data.cumsum()/N

5815


Nurse Practitioner                                                0.134996
Dentist                                                           0.246260
Internal Medicine                                                 0.353912
Family Practice                                                   0.451075
Physician Assistant                                               0.527773
Emergency Medicine                                                0.570765
Student in an Organized Health Care Education/Training Program    0.613758
Obstetrics & Gynecology                                           0.648667
Optometry                                                         0.677558
Psychiatry                                                        0.703697
General Surgery                                                   0.724506
Ophthalmology                                                     0.744282
Cardiology                                                        0.760791
Orthopedic Surgery       

In [30]:
data = part_d['specialty_description'].value_counts()
data = data.cumsum()/N
mask = data < 0.75
mask

value = part_d['specialty_description'].value_counts()
value[mask]

Nurse Practitioner                                                785
Dentist                                                           647
Internal Medicine                                                 626
Family Practice                                                   565
Physician Assistant                                               446
Emergency Medicine                                                250
Student in an Organized Health Care Education/Training Program    250
Obstetrics & Gynecology                                           203
Optometry                                                         168
Psychiatry                                                        152
General Surgery                                                   121
Ophthalmology                                                     115
Name: specialty_description, dtype: int64

## Question 7:
Take a look at these two columns: `total_claim_count` and `opioid_claim_count`. Define a new column that stores the ratio of the two, eg:
```python
part_d['opioid_claim_rate'] = part_d['opioid_claim_count']/part_d['total_claim_count']
```
Let us define **high opioid providers** as those who have an `opioid_claim_rate` that is greater than or equal to 50%. This question has three parts:
1. Run the describe method on the `opioid_claim_rate` column and note the statistics.
2. Use TWO different methods to calculate how many **high opioid providers** there are in the dataset. (Hint: you can use the sum of a boolean variable for one method and the size of a filtered dataset for the other method).
3. Use TWO different methods to identify the top 3 states that have the highest number of **high opioid providers**. (Hint: you can use value_counts for one method and groupby for the other method). 

In [20]:
part_d['opioid_claim_rate'] = part_d['opioid_claim_count']/part_d['total_claim_count']
part_d['opioid_claim_rate']

0       0.036928
1            NaN
2       0.057234
3       0.107958
4            NaN
          ...   
5810    0.033501
5811    0.000000
5812         NaN
5813         NaN
5814    0.024896
Name: opioid_claim_rate, Length: 5815, dtype: float64

In [21]:
#question 7-2 method 1
mask = part_d['opioid_claim_rate']  >= 0.50
print(sum(mask))
#question 7-2 method 2
data = part_d['opioid_claim_rate']
data_high = data[data >=0.50 ]
print(len(data_high))

#question 7-3 method 1
#part_d_high = part_d[mask]
part_d['high_opioid_providers_TF'] = mask
print(part_d.groupby('nppes_provider_state')['high_opioid_providers_TF'].sum().nlargest(3))
#question 7-3 method 2
part_d_high = part_d[part_d['high_opioid_providers_TF'] == True]
part_d_high['nppes_provider_state'].value_counts().head(3)

209
209
nppes_provider_state
CA    19.0
FL    15.0
MI    12.0
Name: high_opioid_providers_TF, dtype: float64


CA    19
FL    15
MI    12
Name: nppes_provider_state, dtype: int64

## Question 8:
Using the `opioid_claim_rate` column computed in the previous question:

1. Create a new DataFrame `g` that aggregates the `nppes_provider_gender` column using the `groupby` method and the `size` function to output the gender breakdown in the dataset. You can use this code to answer this question:
```python
g = part_d.groupby('nppes_provider_gender').size().reset_index()
```
2. Use the `rename` method to rename the columns of DataFrame `g` to *gender* and *count*.
3. Now compute the aggregation of the dataset by gender but filtered to the **high opioid providers**. Name the DataFrame `g2`. Rename the columns to *gender* and *high_opioid_count*.
4. Join the DataFrames g and g2 using the gender column as the key. The final result should look like this:
```
gender	count	high_opioid_count
	F	2458	62
	M	3357	147
```
5. Compute the ratio of `high_opioid_count` to `count` for each row and note the difference

In [22]:
g = part_d.groupby('nppes_provider_gender').size().reset_index()
g = g.rename(columns={'nppes_provider_gender':'gender', 0:'count'})
print(g)
g2 = part_d[part_d['high_opioid_providers_TF']== True].groupby('nppes_provider_gender').size().reset_index()
g2 = g2.rename(columns={'nppes_provider_gender':'gender', 0:'high_opioid_count'})
print(g2)


  gender  count
0      F   2458
1      M   3357
  gender  high_opioid_count
0      F                 62
1      M                147


In [23]:
g.join(g2['high_opioid_count'],how='right',lsuffix='_left',rsuffix='_right')

gender  count  high_opioid_count
0      F   2458                 62
1      M   3357                147

## Question 9:
The Office of Inspector General ([OIG](https://oig.hhs.gov/)) has the authority to **exclude** individuals and entities from Federally funded health care programs. To be excluded is quite serious, as you can see from their website:

>The OIG maintains a list of all currently excluded individuals and entities called the List of Excluded Individuals/Entities (LEIE). Anyone who hires an individual or entity on the LEIE may be subject to civil monetary penalties (CMP). To avoid CMP liability, health care entities should routinely check the list to ensure that new hires and current employees are not on it.

This list is publicaly available, [here](https://oig.hhs.gov/exclusions/exclusions_list.asp). I have downloaded this file in its entirety (in December 2020) and have stripped it down to the columns we need.

Load the provided file into a DataFrame, named `exclusions`:

```python
exclusions = pd.read_csv('physicians_exclusions_sample3.csv')
```
This question has a number of parts:

1. Print the number of rows and columns in the exclusions dataset.
2. Display the first 6 rows the DataFrame. 
3. Display 3 random rows of the DataFrame using the `sample` method.
4. Use the `nunique` method to determine the number of unique NPIs. Compare this to the size of the dataset - what do you observe?
5. Are there any NULL `NPI` values?
6. What is the most frequent `NPI` value?
7. Apply the `unique` method to the NPI column and convert the outcome to a list of NPIs (hint use the `tolist` method). Store the output in a variable called `npis`. What is the length of `npis`?
8. Apply the `isin` method to `part_d` to tag the rows whose NPI is in the list of excluded NPIs. You can use the following code. How many records are there in `excluded1`?
```python
excluded1 = part_d[part_d['npi'].isin(npis)]
```

In [24]:
exclusions = pd.read_csv('physicians_exclusions_sample3.csv')
print(exclusions.shape)
print(exclusions.head(6))
print(exclusions.sample(3))
print(exclusions.nunique(),'\n It is much smaller than the size of the data set, indicating that many samples have the none NPI')
print(exclusions['NPI'].isnull().shape[0])
npis = exclusions['NPI'].unique().tolist()
print('the length of npis is ',len(npis))
excluded1 = part_d[part_d['npi'].isin(npis)]
excluded1

(69858, 4)
   LASTNAME    FIRSTNAME STATE  NPI
0   AALDERS        NANCY    IL    0
1  AALTONEN     NICKOLAS    WA    0
2     AAMIR     MUHAMMAD    MI    0
3     AARON        ALINA    AZ    0
4     AARON  CHRISTOPHER    PA    0
5     AARON      DEBORAH    TX    0
      LASTNAME   FIRSTNAME STATE  NPI
53373   ROBLAS  ESTERHILDA    CA    0
45376   NEWELL      SHARON    AL    0
5830     BLAKE     RICHARD    MI    0
LASTNAME     28591
FIRSTNAME    11443
STATE           60
NPI           5200
dtype: int64 
 It is much smaller than the size of the data set, indicating that many samples have the none NPI
69858
the length of npis is  5200


npi nppes_provider_last_org_name nppes_provider_first_name  \
3814  1659506111                      FRANCIS                   DESMOND   
4481  1770505612                         YANG                  HARRISON   

     nppes_provider_mi nppes_credentials nppes_provider_gender  \
3814               NaN                MD                     M   
4481                 Y  M.D., FACP, FACC                     M   

     nppes_entity_code   nppes_provider_street1 nppes_provider_street2  \
3814                 I  1801 BLACK RIVER BLVD N                    NaN   
4481                 I    845 MCARTHUR ST STE D                    NaN   

     nppes_provider_city  ...  beneficiary_race_black_count  \
3814                ROME  ...                           NaN   
4481          MANCHESTER  ...                          25.0   

      beneficiary_race_asian_pi_count beneficiary_race_hispanic_count  \
3814                              NaN                            15.0   
4481                              NaN                             NaN   

     beneficiary_race_nat_ind_count beneficiary_race_other_count  \
3814                            NaN                          NaN   
4481                            0.0                          NaN   

     beneficiary_nondual_count beneficiary_dual_count  \
3814                     214.0                  138.0   
4481                     168.0                  295.0   

      beneficiary_average_risk_score  opioid_claim_rate  \
3814                          1.3818           0.089926   
4481                          1.6193           0.119433   

      high_opioid_providers_TF  
3814                     False  
4481                     False  

[2 rows x 86 columns]

## Question 10
In question 9 we searched for records in the part_d DataFrame whose NPI was in the exclusions dataset. Matching across the NPI yielded only 2 excluded providers. This seems suspicious. The problem is probably due to the fact that so many of the NPIs in the excluded dataset are marked as `0`. Perhaps we can do better by matching across first and last names. Let's do that in these steps.
1. As a sanity check, let's first do a merge between the two datasets on NPI. Your results should confirm the result obtained in question 9. For convenience, define this list of columns:
```python
cols = [
    'npi', 
    'nppes_provider_first_name', 
    'nppes_provider_last_org_name', 
    'nppes_provider_state', 
    'NPI',
    'FIRSTNAME',
    'LASTNAME',
    'STATE',
    'opioid_claim_rate'
]
```
Perform the merge and display the output for the columns in `cols`. You can use the following code to answer this question:
```python
excluded2 = pd.merge(left = part_d, right = exclusions, how = 'inner', 
         left_on = ['npi'],
         right_on = ['NPI']
        )
excluded2[cols]
```
Your results should match the results from Question 9.


2. Since we are going to match across names, let's first convert the applicable columns to upper case and strip out any leading or trailing spaces. You can use the following code to answer this question but you'll need to apply it to the remaining first and last name columns, eg 'nppes_provider_first_name', 'nppes_provider_last_org_name', 'FIRSTNAME' and 'LASTNAME'
```python
part_d['nppes_provider_first_name'] = part_d['nppes_provider_first_name'].str.upper().str.strip()
```

3. Run another inner join across both datasets using first and last names. My result had 461 rows. Please take a look at the dataset filtering on the columns in `cols` as done above. What do you notice? 


4. In order to conduct a better quality match, include the state in your join keys. You should now be matching on first name, last name and state. My result had 22 records. I feel more confident in this match since the names and state lines up. You'll notice that some of the providers have 0 `opioid_claim_rate`. This is not that surprising as there are many different kinds of behaviors that a provider can engage in in order to get on the exclusisons list. 


In [32]:
cols = [
    'npi', 
    'nppes_provider_first_name', 
    'nppes_provider_last_org_name', 
    'nppes_provider_state', 
    'NPI',
    'FIRSTNAME',
    'LASTNAME',
    'STATE',
    'opioid_claim_rate'
]
excluded2 = pd.merge(left = part_d, right = exclusions, how = 'inner', 
      left_on = ['npi'],
      right_on = ['NPI']
     )
excluded2[cols]

npi nppes_provider_first_name nppes_provider_last_org_name  \
0  1659506111                   DESMOND                      FRANCIS   
1  1770505612                  HARRISON                         YANG   

  nppes_provider_state         NPI FIRSTNAME LASTNAME STATE  opioid_claim_rate  
0                   NY  1659506111   DESMOND  FRANCIS    NY           0.089926  
1                   TN  1770505612  HARRISON     YANG    TN           0.119433

In [33]:
part_d['nppes_provider_first_name'] = part_d['nppes_provider_first_name'].str.upper().str.strip()
excluded2.groupby('nppes_provider_first_name')[cols].size()

nppes_provider_first_name
DESMOND     1
HARRISON    1
dtype: int64

In [34]:
part_d.groupby('nppes_provider_first_name')['opioid_claim_rate'].size()

nppes_provider_first_name
AARON         10
AASHINI        1
ABBEY          2
ABBY           1
ABDELKADER     1
              ..
YVONNE         5
ZACHARY        6
ZAKARIA        1
ZEBULUN        1
ZEESHAN        1
Name: opioid_claim_rate, Length: 2143, dtype: int64

In [38]:
merged = pd.merge(part_d,exclusions,how ='inner',
                 left_on = ['nppes_provider_first_name','nppes_provider_last_org_name'],
                 right_on = ['FIRSTNAME','LASTNAME'])
merged.shape

(435, 90)